website/source/guides/load-balancing/nginx.html.md

```diff
@@ -0,0 +1,281 @@
+to scale along with your services.
+
+The main use case for NGINX in this scenario is to distribute incoming HTTP(S)
+and TCP requests from the internet to frontend services that can handle these
+requests. This guide will show you one such example using a demo web
+application.
+
+## Reference Material
+
+- [NGINX][nginx]
+- [Load Balancing Strategies for Consul][lb-strategies]
+
+## Estimated Time to Complete
+
+20 minutes
+
+## Prerequisites
+
+To perform the tasks described in this guide, you need to have a Nomad
+environment with Consul installed. You can use this [repo][terraform-repo] to
+easily provision a sandbox environment. This guide will assume a cluster with
+one server node and three client nodes.
+
+-> **Note:** This guide is for demo purposes and is only using a single
+server node. In a production cluster, 3 or 5 server nodes are recommended.
+
+## Steps
+
+### Step 1: Create a Job for Demo Web App
+
+Create a job for a demo web application and name the file 'webapp.nomad'
+
+```hcl
+job "demo-webapp" {
+  datacenters = ["dc1"]
+
+  group "demo" {
+    count = 3
+
+    task "server" {
+      env {
+        PORT    = "${NOMAD_PORT_http}"
+        NODE_IP = "${NOMAD_IP_http}"
+      }
+
+      driver = "docker"
+
+      config {
+        image = "hashicorp/demo-webapp-lb-guide"
+      }
+
+      resources {
+        network {
+          mbits = 10
+          port  "http"{}
+        }
+      }
+
+      service {
+        name = "demo-webapp"
+        port = "http"
+
+        check {
+          type     = "http"
+          path     = "/"
+          interval = "2s"
+          timeout  = "2s"
+        }
+      }
+    }
+  }
+}
+```
+
+Note that this job deploys 3 instances of our demo web application which we will
+load balance with NGINX in the next few steps.
+
+### Step 2: Deploy the Demo Web App
+
+We can now deploy our demo web application:
+
+```shell
+$ nomad run webapp.nomad 
+==> Monitoring evaluation "ea1e8528"
+    Evaluation triggered by job "demo-webapp"
+    Allocation "9b4bac9f" created: node "e4637e03", group "demo"
+    Allocation "c386de2d" created: node "983a64df", group "demo"
+    Allocation "082653f0" created: node "f5fdf017", group "demo"
+    Evaluation status changed: "pending" -> "complete"
+==> Evaluation "ea1e8528" finished with status "complete"
+```
+
+### Step 3: Create a Job for NGINX
+
+Create a job for NGINX and name it `nginx.nomad`. This will be our load balancer
+that will balance requests to the deployed instances of our web application.
+
+```hcl
+job "nginx" {
+  datacenters = ["dc1"]
+
+  group "nginx" {
+    count = 1
+
+    task "nginx" {
+      driver = "docker"
+
+      config {
+        image = "nginx"
+
+        port_map {
+          http = 80
+        }
+
+        volumes = [
+          "local:/etc/nginx/conf.d",
+        ]
+      }
+
+      template {
+        data = <<EOF
+upstream backend {
+{{ range service "demo-webapp" }}
+  server {{ .Address }}:{{ .Port }};
```

Comment:
When there are no services for "demo-webapp" in Consul this will result in an invalid Nginx template and the container will crash. Nginx can't handle empty upstreams. As a workaround add:
```
{{else}}server 127.0.0.1:65535; # force a 502
{{end}}
```
See https://github.com/hashicorp/consul-template/blob/master/examples/nginx.md

Reply:
Excellent. Thank you!